### PR TITLE
Implements SeekParameters.*_SYNC variants for HLS

### PR DESCRIPTION
**Pull request by stevemayhew**
The HLS implementation of `getAdjustedSeekPositionUs()` now completely supports `SeekParameters.CLOSEST_SYNC` and it's brotheran, assuming the HLS stream indicates segments all start with an IDR (that is EXT-X-INDEPENDENT-SEGMENTS  is specified).

This fixes issue #2882 and improves (but does not completely solve #8592

### CODE REVIEW

**Comment by stevemayhew** [comment]
@ojw28  Here is the first step to solving the faster first render requested in #8592.   In my testing on BRCM and AmLogic platforms it does not completely solve the issues, but we are looking at ways to intimidate the decoder into presenting a single IDR immediately.

At any rate, here's a fix for one bug on your plate :-).    I am happy to split out the test cases into a set that test the logic in `SeekParameters.resolveSeekPositionUs()` as a separate unit-test (seemed more proper way),  I did it this way to limit the change to a single module.   It could also be tech-debt I'm happy to take on.

Still working on a design for the fix for #8952.  Haven't forgotten about this one.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java [hunk 1]
```diff
@@ -237,6 +238,31 @@ public void setIsTimestampMaster(boolean isTimestampMaster) {
...
+    long adjustedPositionUs = positionUs;
+
+    int selectedIndex = trackSelection.getSelectedIndex();
+    boolean haveTrackSelection = selectedIndex < playlistUrls.length && selectedIndex != C.INDEX_UNSET;
+    @Nullable HlsMediaPlaylist mediaPlaylist = null;
+    if (haveTrackSelection) {
+      mediaPlaylist = playlistTracker.getPlaylistSnapshot(playlistUrls[selectedIndex], /* isForPlayback= */ true);
+    }
+
+    // Resolve to a segment boundary, current track is fine (all should be same).
```
Comment (ojw28) [inline library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: > current track is fine (all should be same)

I don't think this is required by the HLS spec, and we've definitely seen cases where streams deliberately use different segment durations for different playlists, in which case the segment boundaries are no longer aligned.

Presumably it's highly likely that the seek will be performed into the current track, and making this work for the common case definitely seems worth it for possibly getting things wrong in the edge case that there's a track change _and_ segments are not aligned. So this is still fine, but maybe delete the comment?

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java [hunk 2]
```diff
@@ -237,6 +238,31 @@ public void setIsTimestampMaster(boolean isTimestampMaster) {
...
+    // and, segments must start with sync (EXT-X-INDEPENDENT-SEGMENTS must be present)
+    if (mediaPlaylist != null && mediaPlaylist.hasIndependentSegments) {
```
Comment (ojw28) [inline library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: You probably need to check `!mediaPlaylist.segments.isEmpty()` as part of this condition, to avoid making any adjustment in the edge case where there aren't any segments in the playlist yet.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java [hunk 3]
```diff
@@ -237,6 +238,31 @@ public void setIsTimestampMaster(boolean isTimestampMaster) {
...
+      int segIndex = Util.binarySearchFloor(mediaPlaylist.segments, positionUs, true, true);
```
Comment (ojw28) [inline library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: This is using the absolute `positionUs`, but I think the search is expecting a relative position. So I think you need to do something like:

```
long startOfPlaylistInPeriodUs = playlist.startTimeUs - playlistTracker.getInitialStartTimeUs();
long targetPositionInPlaylistUs = positionUs - startOfPlaylistInPeriodUs;
```

and then use `targetPositionInPlaylistUs` to do the binary search. You'll then need to add `startOfPlaylistInPeriodUs` back into `firstSyncUs` and `secondSyncUs` so that they're absolute positions, before resolving the seek position.

Please double check you agree with this. As a point of reference, look at the other place in this class that calls `Util.binarySearchFloor`, and look at what value it's passing in. I think that my suggestion above is consistent with what's happening there.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java [hunk 4]
```diff
@@ -237,6 +238,31 @@ public void setIsTimestampMaster(boolean isTimestampMaster) {
     this.isTimestampMaster = isTimestampMaster;
   }
 
+  long getAdjustedSeekPositionUs(long positionUs, SeekParameters seekParameters) {
```
Comment (ojw28) [inline library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Please make public, and add Javadoc.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java [hunk 5]
```diff
@@ -585,6 +586,16 @@ public boolean onPlaylistError(Uri playlistUrl, LoadErrorInfo loadErrorInfo, boo
         && exclusionDurationMs != C.TIME_UNSET;
   }
 
+  public boolean isVideoSampleStream() {
```
Comment (ojw28) [inline library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:596]: Add Javadoc. Ditto for the method below.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+        mockExtractorFactory,
+        mockPlaylistTracker,
+        new Uri[] {PLAYLIST_URI},
+        new Format[] { ExoPlayerTestRunner.VIDEO_FORMAT },
+        mockDataSourceFactory,
+        null,
+        new TimestampAdjusterProvider(),
+        null);
+
+    assertThat(testee).isNotNull();
```
Comment (ojw28) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java]: This assertion is guaranteed to be true from the line immediately above. Remove.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+  }
+
+  @Test
+  public void getAdjustedSeekPositionUs_PreviousSync() {
+    when(mockPlaylistTracker.isSnapshotValid(eq(PLAYLIST_URI))).thenReturn(true);
```
Comment (ojw28) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java]: Any reason not to pull this up from every test method into `setup`?

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+  @Mock
+  private HlsDataSourceFactory mockDataSourceFactory;
+  private HlsChunkSource testee;
+
+  @Before
+  public void setup() throws IOException {
+    // sadly, auto mock does not work, you get NoClassDefFoundError: com/android/dx/rop/type/Type
+//    MockitoAnnotations.initMocks(this);
+    mockExtractorFactory = Mockito.mock(HlsExtractorFactory.class);
```
Comment (ojw28) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java]: Why can't you just pass a real instance (i.e., `HlsExtractorFactory.DEFAULT`)? If this works then it's generally much better than mocking. In particular because this test doesn't randomly start failing due to some unrelated change that ends up meaning the mock starts to be called.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    mockPlaylistTracker = Mockito.mock(HlsPlaylistTracker.class);
+    mockDataSourceFactory = Mockito.mock(HlsDataSourceFactory.class);
```
Comment (ojw28) [inline library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java:82]: Ditto. I'm not sure about the `HlsPlaylistTracker`. It's possible that you actually need to mock that, but please take a look.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 1 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Yes, that sounds good.   We are actually working on another bug with how HlsChunkSource switches tracks using MSN to locate the matching segment.  This is explicitly forbidden in the spec when it is intra-variant:

From 6.3.2

Allowed within the same playlist:
>    A client MAY use the segment Media Sequence Number to track the
   location of a Media Segment within a Playlist when the Playlist is
   reloaded.

But not intra-variant:

>    A client MUST NOT assume that segments with the same Media Sequence
   Number in different Variant Streams or Renditions have the same
   position in the presentation; Playlists MAY have independent Media
   Sequence Numbers.  Instead, a client MUST use the relative position
   of each segment on the Playlist timeline and its Discontinuity
   Sequence Number to locate corresponding segments

Where this matters is we have an origin vendor that is using GAP to mark dropped GOP sections from the transcode stream for individual variants.   As such, they may have playlist that look like:

V1:  GAP(2s) 1a - Segment 1b (4s) - Segment 2(6s)
V2: Segment 1 (6s) - Segment 2 (6s)

So Segment 2 in the two variants has the same content, but different MSN.  I believe this will break the logic in `HlsChunkSource.getNextChunk()`

For this particular method, consider the case where the current index playlist is V1, if the time locates "Segment 1b" start as nearest sync.  If the track switch changes to V2 between this call to `getAdjustedSeekPositionUs()` and the load for the `seekTo()` operations then yes we would land in the middle of V2's Segment 1.

Bottom line, remove the comment is a good idea, it shouldn't matter if the segment counts/durations change because everything is playlist time based, which is mandated to be consistent intra-playlist:

> Instead, a client MUST use the relative position
   of each segment on the Playlist timeline and its Discontinuity
   Sequence Number to locate corresponding segments.

How about the test case?  Are you ok with this in HLS or should I move the logic to test `SeekParameters` itself (IMHO this is the better place if we are ok with changes to core library)

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 2 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Good call.  I'll add a test case for this too. `Util.binarySearchFloor()` will return 0 in this case, which will lead to an IndexOutOfBounds exception.   Thanks for finding that bug.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 3 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Yes, I'd like to ask about this...  So is the idea to normalize different staleness for playlist snapshots when you have to switch tracks?  Probably worth adding some unit test case logic to cover this case if it is possible.  If there is no playlist switch then the period relative time vs inter-playlist time shouldn't be different.  Maybe @tonihei or @AquilesCanta have  some insight here.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 4 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Sure, I'll follow what core did for DASH.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java (hunk 5 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsSampleStreamWrapper.java:596]: sure.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java (hunk 8 above)
Comment (stevemayhew) [reply library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java]: I'll give it a try, yes sadly Mockito comes with a host of issues with Android mocks in Roboelectric.  Also, I'm fine with moving most of this to a test of `SeekParameters` directly.  Then it covers all users of that code not use HLS.

Only HLS specific stuff will be empty playlist and if we can test track switching (ABR)

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 3 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Ah, yes.. the `Player.seekTo()` is relative to the `Window` in the `Timeline` but looks like `ExoPlayerImplInternal` resolves this as period time (not clear in the `MediaPeriod` Javadoc, but thinking about it of course all `positionUs` values in that API are in period time).  Certainly hard to follow in `ExoPlayerImplInternal.seekToInternal()`

Will fix and add to unit tests to cover this.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 2 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: Fixed, see `getAdjustedSeekPositionUs_EmptyPlaylist()` test and isEmpty check

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 1 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: I deleted the comment,  we will put a sample stream together with tracks where segment media sequence is not aligned and enter bugs for this if there are playback issues as described above.

File: library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java (hunk 3 above)
Comment (stevemayhew) [reply library/hls/src/main/java/com/google/android/exoplayer2/source/hls/HlsChunkSource.java]: This is fixed in latest version of the commit to use `MediaPeriod` relative `positionUs` and covert it to/from playlist relative.  Test cases updated as well.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java (hunk 9 above)
Comment (stevemayhew) [reply library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java:82]: Yes, using the `DefaultHlsPlaylistTracker`, IMO, would add a lot of complexity (`Loader`, `DataSource`, etc) when all we need it to do is return the current playlist, report it valid and the period start time.

Always worth considering the cost/benefit of mocking an object vs using the defaults though.

**Comment by stevemayhew** [comment]
@ojw28  I think all of the comments/issues have been addressed.   Do let me know, thanks so much for the review and catching the bug with `MediaPeriod` vs `Timeline` / playlist relative time.

File: library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java (hunk 7 above)
Comment (stevemayhew) [reply library/hls/src/test/java/com/google/android/exoplayer2/source/hls/HlsChunkSourceTest.java]: done

**Comment by ojw28** [comment]
@stevemayhew - Are you certain you pushed all of your changes? Looking at your comment starting "I deleted the comment", it appears the comment is still there. That's not particularly important in itself, but it's very hard to tell whether there are also other changes that haven't been pushed?

I'm not sure if you got into a bad state and were forced into it, but force pushing a single commit on top of a previous review makes it incredibly hard to work out what changed. Please just push additional commits, if possible.

**Comment by stevemayhew** [comment]
@ojw28  Yes I updated the commit and did a force-push.   Sure, in the future I'll make new commits each time for reviews you are on.   Some people don't like lots of commits on a pull request branch, but I can see addressing review comments it is nice to see the diffs.

I do like when people interactive rebase to a single change before the final merge, that makes a cleaner history.  But next time, I'll have you let me know when you are ready to merge then I can do that before the final merge.

Thanks for the review...  BTW, the change does speed up paused seeks on i-Frame only playlist a bit.  I've got some other things I'm working on for #8592 as well.

Once it is all a bit more baked I can add a track selector setting for iFrame only mode and add the scrub seek to the demo app.

**Comment by ojw28** [comment]
So is the comment not being deleted just an omission? I guess I saw that along with the "done" comment, and was then unsure whether there are other accidental omissions too (and can't easily see a diff to check). If you could confirm that everything is as it should be, and that the comment not being deleted was just a singular omission, that would be great. Thanks!

**Comment by ojw28** [comment]
In case it's not completely clear what I'm referring to, I'm referring to this (comment in the code is still there, but comment in the discussion thread says the comment in the code has been deleted):

https://github.com/google/ExoPlayer/pull/9536/files#diff-a500e9307d881d3b7b6a92999132c39808322d24b6f6a4a57a1c0d03cd6cf513R260

**Comment by stevemayhew** [comment]
Let me double check it all over and diff against our version (which is very close, but based from release 2.12) just to be double sure.

The comment was not removed, it is misleading.

The resolved sync point position will be valid in any variant, however it may not match a segment boundary. If for example there were sub-segment duration `EXT-X-GAP` segments in only one playlist such as to cause an disparate list of segments inter-variant.

This actually will cause the MSN based algorithm later in this class to pick in incorrect segment on track switch, so probably better just to add a bug

**Comment by stevemayhew** [comment]
@ojw28 I double checked all the diffs match up with what we have in our 2.12 version.   The unit test cases all run.

I updated the comment so it is correct now,  or I can remove it completely too if that is preferred.   Also, let me know if  you want me to squash the commits into one once we agree the review is complete and it is ready to merge.   Certainly makes it easier to merge then bunches of edit commits from review comments.  In the future I will not do this until it is ready to merge.   Sorry again for the confusion.

**Comment by ojw28** [comment]
Thanks! We'll get this merged.

> Also, let me know if you want me to squash the commits into one once we agree the review is complete and it is ready to merge. Certainly makes it easier to merge then bunches of edit commits from review comments.

The tooling & internal review process we use gives us with a flat view anyway, so it doesn't really make any difference to us. The only benefit of squashing is that it makes the commit history look a little neater in the end. On the flip side, anyone wanting to revisit the review of this PR later and understand any of the discussion threads will find it much harder if there's a squash and force push, because some of those threads wont be tied to any accessible source code any more. I think being able to revisit the PR review is probably more important, so overall I think that not squashing is probably preferable for cases like this.
